parallel.go

```diff
@@ -0,0 +1,284 @@
+			newIdx,
+		}
+		(*h)[0] = newKey
+		heap.Fix(h, 0)
+	} else {
+		heap.Pop(h)
+	}
+	return k
+}
+
+func (h *bitmapContainerHeap) PopNextContainers() multipleContainers {
+	if h.Len() == 0 {
+		return multipleContainers{}
+	}
+
+	containers := make([]container, 0, 4)
+	bk := h.PopIncrementing()
+	containers = append(containers, bk.container)
+	key := bk.key
+
+	for h.Len() > 0 && key == h.Peek().key {
+		bk = h.PopIncrementing()
+		containers = append(containers, bk.container)
+	}
+
+	return multipleContainers{
+		key,
+		containers,
+		-1,
+	}
+}
+
+func newBitmapContainerHeap(bitmaps ...*Bitmap) bitmapContainerHeap {
+	// Initialize heap
+	var h bitmapContainerHeap = make([]bitmapContainerKey, 0, len(bitmaps))
+	for _, bitmap := range bitmaps {
+		if !bitmap.IsEmpty() {
+			key := bitmapContainerKey{
+				bitmap,
+				bitmap.highlowcontainer.getWritableContainerAtIndex(0),
+				bitmap.highlowcontainer.keys[0],
+				0,
+			}
+			h = append(h, key)
+		}
+	}
+
+	heap.Init(&h)
+
+	return h
+}
+
+func repairAfterLazy(c container) container {
+	switch t := c.(type) {
+	case *bitmapContainer:
+		if t.cardinality == invalidCardinality {
+			t.computeCardinality()
+		}
+
+		if t.getCardinality() <= arrayDefaultMaxSize {
+			return t.toArrayContainer()
+		} else if c.(*bitmapContainer).isFull() {
+			return newRunContainer16Range(0, MaxUint16)
+		}
+	}
+
+	return c
+}
+
+func toBitmapContainer(c container) container {
+	switch t := c.(type) {
+	case *arrayContainer:
+		return t.toBitmapContainer()
+	case *runContainer16:
+		if !t.isFull() {
+			return t.toBitmapContainer()
+		}
+	}
+	return c
+}
+
+func appenderRoutine(bitmapChan chan<- *Bitmap, resultChan <-chan keyedContainer, expectedKeysChan <-chan int) {
+	expectedKeys := -1
+	appendedKeys := 0
+	keys := make([]uint16, 0)
+	containers := make([]container, 0)
+	for appendedKeys != expectedKeys {
+		select {
+		case item := <-resultChan:
+			if len(keys) <= item.idx {
+				keys = append(keys, make([]uint16, item.idx-len(keys)+1)...)
+				containers = append(containers, make([]container, item.idx-len(containers)+1)...)
+			}
+			keys[item.idx] = item.key
+			containers[item.idx] = item.container
+
+			appendedKeys += 1
+		case msg := <-expectedKeysChan:
+			expectedKeys = msg
+		}
+	}
+	answer := &Bitmap{
+		roaringArray{
+			make([]uint16, 0, expectedKeys),
+			make([]container, 0, expectedKeys),
+			make([]bool, 0, expectedKeys),
+			false,
+			nil,
+		},
+	}
+	for i := range keys {
+		answer.highlowcontainer.appendContainer(keys[i], containers[i], false)
+	}
+
+	bitmapChan <- answer
+}
+
+func ParOr(bitmaps ...*Bitmap) *Bitmap {
+	h := newBitmapContainerHeap(bitmaps...)
+
+	bitmapChan := make(chan *Bitmap)
+	inputChan := make(chan multipleContainers, 128)
+	resultChan := make(chan keyedContainer, 32)
+	expectedKeysChan := make(chan int)
+
+	orFunc := func() {
+		for input := range inputChan {
+			c := toBitmapContainer(input.containers[0]).lazyOR(input.containers[1])
```

Comment:
This seems to assume that input.containers.length >= 2. Is this enforced/documented?

Reply:
Yes, that's correct. If count is one then we can pass the resulting bitmap straight to the _appender_ goroutine: as seen around [line 211](https://github.com/RoaringBitmap/roaring/pull/109/files#diff-283cb5309c30ff97b95802636f247453R211).

Reply:
Can you add a test where you do ParOr(rb) where rb is a single bitmap, just so that things are clear?

Reply:
That's very good point. The code actually failed for a one-bitmap case in `ParAnd`. It's now fixed.